source/synthDrivers/espeak.py

```diff
@@ -90,8 +99,69 @@ def _processText(self, text):
 			0x5B: u" [", # [: [[ indicates phonemes
 		})
 
-	def speak(self,speechSequence):
-		defaultLanguage=self._language
+	def _determineLangFromCommand(self, command: LangChangeCommand) -> Optional[str]:
+		"""
+		Checks if a lang code given is compatible with eSpeak.
+		If not, check if a default mapping occurs in L{_defaultLangToLocale}.
+		Otherwise, checks if a language of a different dialect exists (e.g. ru-ru to ru).
+		Returns an eSpeak compatible lang code or None, which should be handled as
+		falling back to the default language.
+		"""
+		# Use default language if no command.lang is supplied
+		langWithLocale = command.lang if command.lang else self._language
+		langWithLocale = langWithLocale.lower().replace('_', '-')
+
+		langWithoutLocale: Optional[str] = langWithLocale.split('-')[0]
+
+		# Check for any language where the language code matches, regardless of dialect: e.g. ru-ru to ru
+		matchingLanguages = filter(lambda l: l.split('-')[0] == langWithoutLocale, self.availableLanguages)
+		anyLocaleMatchingLang = next(matchingLanguages, None)
+		
+		# Check from a list of known default mapping locales: e.g. en to en-gb
+		# Created due to eSpeak issue: https://github.com/espeak-ng/espeak-ng/issues/1200
+		knownDefaultLang = self._defaultLangToLocale.get(langWithoutLocale, None)
+		if knownDefaultLang is not None and knownDefaultLang not in self.availableLanguages:
+			# This means eSpeak has changed and we need to update the mapping
+			log.error(f"Default mapping unknown to eSpeak {knownDefaultLang} not in {self.availableLanguages}")
+			knownDefaultLang = None
+
+		if langWithLocale in self.availableLanguages:
+			eSpeakLang = langWithLocale
+		elif knownDefaultLang is not None:
+			eSpeakLang = knownDefaultLang
+		elif langWithoutLocale in self.availableLanguages:
+			eSpeakLang = langWithoutLocale
+		elif anyLocaleMatchingLang is not None:
+			eSpeakLang = anyLocaleMatchingLang
+		else:
+			log.debugWarning(f"Unable to find an eSpeak language for '{langWithLocale}'")
+			eSpeakLang = None
+		return eSpeakLang
+
+	def _handleLangChangeCommand(
+			self,
+			langChangeCommand: LangChangeCommand,
+			textList: List[str],
+			langChanged: bool,
+	) -> bool:
+		"""Append language xml tags to textList:
+			- if a language change has already been handled for this speech,
+			close the open voice tag.
+			- if the language is supported by eSpeak, switch to that language.
+			- otherwise, switch to the default synthesizer language.
+		"""
+		lang = self._determineLangFromCommand(langChangeCommand)
+		if langChanged:
+			textList.append("</voice>")
+		if lang is not None:
+			textList.append(f'<voice xml:lang="{lang}">')
+		else:
+			textList.append(f'<voice>')

```

Comment:
In two cases the result of `self._determineLangFromCommand` isn't used. I think this logic would be clearer in the outer scope.

Reply:
This seems to contradict with https://github.com/nvaccess/nvda/pull/13739#discussion_r888692553.
Let me know if this isn't resolved by the fix for that review comment in 5c3388b